File: deepspeed/inference/v2/kernels/ragged_ops/logits_gather/logits_gather.cpp
// Copyright (c) Microsoft Corporation.
// SPDX-License-Identifier: Apache-2.0

// DeepSpeed Team

#include "logits_gather.h"

#define DISPATCH_TO_LOGITS_GATHER(T_TYPE, C_TYPE)                  \
    if (all_acts.options().dtype() == torch::T_TYPE) {             \
        launch_logits_gather((C_TYPE*)final_token_acts.data_ptr(), \
                             (const C_TYPE*)all_acts.data_ptr(),   \
                             batch_metadata_raw,                   \
                             seq_metadata_raw,                     \
                             n_seqs,                               \
                             embed_dim,                            \
                             at::cuda::getCurrentCUDAStream());    \
    }

/*
Logits gather will parse the ragged batch data structure and gather only the logits that
will be used for token sampling.
*/
void gather_for_logits(torch::Tensor& final_token_acts,
                       torch::Tensor& all_acts,
                       torch::Tensor& batch_metadata,
                       torch::Tensor& seq_metadata)
{
    const RaggedBatchDescriptor* batch_metadata_raw =
        reinterpret_cast<const RaggedBatchDescriptor*>(batch_metadata.data_ptr());

    const InflightSeqDescriptor* seq_metadata_raw =
        reinterpret_cast<const InflightSeqDescriptor*>(seq_metadata.data_ptr());

    const int n_seqs = final_token_acts.size(0);
    const int embed_dim = final_token_acts.size(1);

    TORCH_CHECK(all_acts.scalar_type() == final_token_acts.scalar_type(),
                "all_acts and final_token_acts must have the same scalar type");

    DISPATCH_TO_LOGITS_GATHER(kFloat, float)
    DISPATCH_TO_LOGITS_GATHER(kHalf, half)
#ifdef BF16_AVAILABLE
    DISPATCH_TO_LOGITS_GATHER(kBFloat16, __nv_bfloat16)
#endif
}

File: deepspeed/inference/v2/kernels/ragged_ops/logits_gather/logits_gather.py
# Copyright (c) Microsoft Corporation.
# SPDX-License-Identifier: Apache-2.0

# DeepSpeed Team

import torch

from ... import DSKernelBase
from deepspeed.ops.op_builder import RaggedOpsBuilder
from ....inference_utils import elem_size
from ....ragged import RaggedBatchWrapper


class RaggedLogitsGather(DSKernelBase):
    """
    CUDA Kernel implementation for gather the hidden states of the final token
    of each sequence. This is used to reduce the cost of the performing the unembedding.
    """

    supported_dtypes = [torch.float16, torch.bfloat16, torch.float32]

    def __init__(self, model_dim: int, fp_dtype: torch.dtype):
        """
        Parameters:
            fp_dtype (torch.dtype): Data type for the input/output. Supported values
                are torch.float16, torch.bfloat16, and torch.float32.
        """
        if fp_dtype not in RaggedLogitsGather.supported_dtypes:
            raise ValueError("Unsupported data type: {}, supported_dtypes are {}".format(
                fp_dtype, RaggedLogitsGather.supported_dtypes))

        if elem_size(fp_dtype) * model_dim % 16 != 0:
            raise ValueError("Embedding dimension must be aligned to 16 bytes, got {}".format(model_dim))

        inf_module = RaggedOpsBuilder().load()
        self.kernel = inf_module.gather_for_logits

    def __call__(self, final_token_activations: torch.Tensor, all_activations: torch.Tensor,
                 ragged_wrapper: RaggedBatchWrapper) -> torch.Tensor:
        """
        Gather the hidden states of the final token of each sequence from `all_activations` into
        `final_token_activations`.

        Args:
            final_token_activations (torch.Tensor): Output tensor of shape [num_seqs, model_dim]
            all_activations (torch.Tensor): Input tensor of shape [num_tokens, model_dim]
            ragged_wrapper (RaggedBatchWrapper): Wrapper for the ragged batch.
        """

        self.kernel(final_token_activations, all_activations, ragged_wrapper.batch_metadata_buffer(),
                    ragged_wrapper.inflight_seq_descriptors())
        return final_token_activations
